### PR TITLE
[FEATURE] Filter data of layers by polygon depending on the user groups

### PR DESCRIPTION
**Pull request by mdouchin**
This is a new feature, which requires:

* Lizmap plugin >= 3.6.0.
* QGIS Server >= 3.4.0, but is optimized for QGIS 3.16 and above

It allows to filter the layers data spatially by testing the intersection
of the features against the features of the configured polygon layer.

The filtering polygons are selected for the authenticated user based on a chosen field containing a list of user groups.
* At present, if the user is not authenticated, the filtered layers will show no data at all.
* The filter can be deactivated for some groups of users with the already existing right
`lizmap.tools.loginFilteredLayers.override` labeled "See all the data of the filtered layers (attribute or spatial filters)"

* Funded by Valabre, Centre de gravité de la formation des métiers de la Sécurité Civile, de la Recherche, des Nouvelles Technologies et de la Prévention dans le domaine des risques naturels. htps://www.valabre.com

### CODE REVIEW

**Comment by mdouchin** [comment]
Nothing yet, just a basic QGIS project with 2 layers, for manual and/or automatic tests

**Comment by Gustry** [comment]
Requested linked to https://github.com/3liz/lizmap-plugin/pull/347

File: tests/qgis-projects/tests/test_filter_layer_data_by_polygon_for_groups.qgs [hunk 1]
```diff
@@ -0,0 +1,1310 @@
+<qgis projectname="" saveDateTime="2021-06-22T17:54:09" saveUser="mdouchin" saveUserFull="mdouchin" version="3.16.5-Hannover">
```
Comment (Gustry) [inline tests/qgis-projects/tests/test_filter_layer_data_by_polygon_for_groups.qgs:1]: The project needs to be updated according to what was done in https://github.com/3liz/lizmap-plugin/pull/355

File: tests/qgis-projects/tests/test_filter_layer_data_by_polygon_for_groups.qgs.cfg [hunk 2]
```diff
@@ -0,0 +1,128 @@
+{
```
Comment (Gustry) [inline tests/qgis-projects/tests/test_filter_layer_data_by_polygon_for_groups.qgs.cfg:1]: This CFG needs to updated from https://github.com/3liz/lizmap-plugin/pull/355

The configuration was still missing at that time.

File: tests/qgis-projects/tests/test_filter_layer_data_by_polygon_for_groups/lizmap_polygon_filter.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
+# Headers and parameters
```
Comment (Gustry) [inline tests/qgis-projects/tests/test_filter_layer_data_by_polygon_for_groups/lizmap_polygon_filter.py:1]: This file can be removed, integrated in https://github.com/3liz/lizmap-plugin/pull/355

**Review by Gustry — COMMENTED** [review]
The PR https://github.com/3liz/lizmap-plugin/pull/355 is ready for some testing with this PR.

There is still a small todo on the server side.

**Comment by mdouchin** [comment]
New commit force pushed. Todo:

* test with Cypress
* create an issue to fix the error with popup editing capabilities not well reported in the context of multiple editable layers
* choose if we must support this filter without the Lizmap plugin server-side (or with a Lizmap plugin older than the next 3.6.0)
* Check with heavy data (1 000 000 parcels polygons filtered with 100 city polygons) and see if the cache must be used (and improved with the new centralized project caching system
* Test with the print

**Comment by mdouchin** [comment]
@laurentj @rldhont @nboisteault 

I just removed the draft flag. What needs to be done, with your kind help ;-)

* @Gustry will merge the related lizmap plugin PR to the master branch (so that you will be able to test the plugin for QGIS Server) 
* @nboisteault will help me to add some tests with Cypress. 
* @laurentj -> I will need some help to understand how to fix the PHP unit tests

Thanks in advance

**Comment by nboisteault** [comment]
@mdouchin users in `admins` can view every layers only if `lizmap.tools.loginFilteredLayers.override` is set to `true` right ?

**Comment by nboisteault** [comment]
@mdouchin I have added tests. I've not tested when logged as admin as I think `lizmap.tools.loginFilteredLayers.override` should be tested apart, but tell me if you really need it for this PR.

**Comment by mdouchin** [comment]
Good for me, thanks a lot.

File: lizmap/modules/lizmap/classes/qgisVectorLayer.class.php [hunk 4]
```diff
@@ -1084,4 +1241,170 @@ public function getRealEditionCapabilities()
...
+        );
+
+        // No filter if Lizmap plugin is not installed server side
+        $plugins = $this->project->getQgisServerPlugins();
+        if (!array_key_exists('Lizmap', $plugins)) {
+            return $no_data_array;
+        }
+
+        // Check if the current user is connected
+        $appContext = \lizmap::getAppContext();
```
[the remark this answers is not in the document]
Comment (laurentj) [reply lizmap/modules/lizmap/classes/qgisVectorLayer.class.php:1288]: Forget for the moment. I will fix it in a futur patch, there are many other place where we could optimize this thing.

**Review by laurentj — APPROVED** [review]
Ok for me for the integration with other classes (but no review for the SIG part)

**Comment by mdouchin** [comment]
Thanks @laurentj 
@rldhont can you also review this PR when you have time ?

**Review by rldhont — APPROVED** [review]

**Comment by mdouchin** [comment]
I just rebased some commits and pushed (-f). I am merging right now. Thanks a lot for all participants !

**Pull request merged**